File: src/main/java/io/github/arieslab/pages/ByClassTestPage.java
package io.github.arieslab.pages;

import io.github.arieslab.WicketApplication;
import io.github.arieslab.business.ProjetoBusiness;
import io.github.arieslab.base.Util;
import io.github.arieslab.dtolocal.TotalProcessado;
import io.github.arieslab.base.JNose;
import io.github.arieslab.dtolocal.ProjetoDTO;
import io.github.arieslab.entities.Projeto;
import io.github.arieslab.pages.base.BasePage;
import com.googlecode.wicket.jquery.ui.panel.JQueryFeedbackPanel;
import org.apache.wicket.AttributeModifier;
import org.apache.wicket.ajax.AbstractAjaxTimerBehavior;
import org.apache.wicket.ajax.AjaxRequestTarget;
import org.apache.wicket.ajax.markup.html.form.AjaxCheckBox;
import org.apache.wicket.extensions.ajax.markup.html.IndicatingAjaxLink;
import org.apache.wicket.markup.html.WebMarkupContainer;
import org.apache.wicket.markup.html.link.Link;
import org.apache.wicket.markup.html.list.ListItem;
import org.apache.wicket.markup.html.list.ListView;
import org.apache.wicket.markup.html.panel.FeedbackPanel;
import org.apache.wicket.model.Model;
import org.apache.wicket.markup.html.basic.Label;
import org.apache.wicket.model.PropertyModel;
import org.apache.wicket.protocol.http.WebApplication;
import org.apache.wicket.spring.injection.annot.SpringBean;


import java.io.*;
import java.util.*;
import java.util.List;

import static java.lang.System.out;

/**
 * Page for processing projects grouped by test class, with optional coverage analysis.
 */
public class ByClassTestPage extends BasePage {
    private static final long serialVersionUID = 1L;

    private String pastaPath;
    private String pathAppToWebapp;
    private String pastaPathReport;
    private List<ProjetoDTO> listaProjetos;
    private ListView<ProjetoDTO> lvProjetos;
    private Label taLog;
    private TotalProcessado totalProcessado;
    private Boolean processando;
    private IndicatingAjaxLink processarTodos;
    private StringBuffer logRetorno;
    private String dataProcessamentoAtual;
    private boolean processarCobertura;
    private List<List<String>> listaResultado;

    private Link lkResultadoBotton;

    @SpringBean
    private ProjetoBusiness projetoBusiness;

    public ByClassTestPage() {
        super("ByClassTestPage");

        pathAppToWebapp = WebApplication.get().getServletContext().getRealPath("");
        pastaPathReport = pathAppToWebapp + File.separatorChar + "reports" + File.separatorChar;
        pastaPath = "";
        logRetorno = new StringBuffer();
        processando = false;
        processarCobertura = false;
        totalProcessado = new TotalProcessado();
        totalProcessado.setValor(0);

        criarCheckBoxCobertura();

        taLog = new Label("taLog");
        taLog.setEscapeModelStrings(false).setOutputMarkupId(true).setOutputMarkupPlaceholderTag(true);
        add(taLog);

        criarListaProjetos();

        FeedbackPanel feedback = new JQueryFeedbackPanel("feedback");
        add(feedback.setOutputMarkupId(true));

        criarBotaoProcessarTodos();

        lkResultadoBotton = new Link<String>("lkResultado") {
            @Override
            public void onClick() {
                setResponsePage(new ResultPage(listaResultado, "Result By ClassTest", "result_byclasstest", true));
            }
        };
        lkResultadoBotton.setEnabled(processando);
        lkResultadoBotton.setOutputMarkupId(true);
        lkResultadoBotton.setOutputMarkupPlaceholderTag(true);
        add(lkResultadoBotton);

        Link lkCharts = new Link<String>("lkCharts") {
            @Override
            public void onClick() {
            }
        };
        add(lkCharts.setVisible(false));

        loadProjetos();

        criarTimer();
    }

    /**
     * Creates the "Process All" button that triggers test smell and optional coverage analysis.
     */
    private void criarBotaoProcessarTodos() {
        processarTodos = new IndicatingAjaxLink<String>("processarTodos") {
            @Override
            public void onClick(AjaxRequestTarget target) {
                processando = true;
                List<ProjetoDTO> listaParaProcessar = new ArrayList<>();

                for (ProjetoDTO projetoDTO : listaProjetos) {
                    if (projetoDTO.getParaProcessar()) {
                        listaParaProcessar.add(projetoDTO);
                    }
                }
                listaResultado = JNose.processarProjetos2(listaParaProcessar, dataProcessamentoAtual, totalProcessado, pastaPathReport, logRetorno);
            }
        };
        processarTodos.setEnabled(false);
        add(processarTodos);
    }

    /**
     * Creates the checkbox for enabling/disabling coverage analysis.
     */
    private void criarCheckBoxCobertura() {
        AjaxCheckBox acbCobertura = new AjaxCheckBox("acbCobertura", new PropertyModel(this, "processarCobertura")) {
            @Override
            protected void onUpdate(AjaxRequestTarget ajaxRequestTarget) {
                WicketApplication.COBERTURA_ON = processarCobertura;
                out.println("COVERAGE_ON: " + processarCobertura);
                logRetorno.insert(0, "COVERAGE_ON: " + processarCobertura + " <br>");
            }
        };
        add(acbCobertura);
    }


    /**
     * Loads all projects from the database into the page.
     */
    private void loadProjetos() {
        dataProcessamentoAtual = Util.dateNowFolder();
        logRetorno = new StringBuffer();
        totalProcessado.setValor(0);
        List<Projeto> listaProjetosBean = projetoBusiness.listAllWithFilter();
        for (Projeto projeto : listaProjetosBean) {
            listaProjetos.add(new ProjetoDTO(projeto));
        }
        lvProjetos.setList(listaProjetos);
        processarTodos.setEnabled(true);
    }

    /**
     * Creates the project list view with selection, progress, and result links.
     */
    private void criarListaProjetos() {
        listaProjetos = new ArrayList<>();

        lvProjetos = new ListView<ProjetoDTO>("lvProjetos", listaProjetos) {
            @Override
            protected void populateItem(ListItem<ProjetoDTO> item) {

                ProjetoDTO projetoDTO = item.getModelObject();

                Link lkCharts = new Link<String>("lkCharts") {
                    @Override
                    public void onClick() {
                    }
                };
                lkCharts.setEnabled(projetoDTO.getProcessado());
                lkCharts.setOutputMarkupId(true);
                lkCharts.setOutputMarkupPlaceholderTag(true);
                item.add(lkCharts);
                projetoDTO.lkCharts = lkCharts;

                Link lkResultado = new Link<String>("lkResultado") {
                    @Override
                    public void onClick() {
                        setResponsePage(new ResultPage(projetoDTO.getResultado(), "Result By ClassTest: " + projetoDTO.getName(), projetoDTO.getName() + "_result_byclasstest", true));
                    }
                };
                lkResultado.setEnabled(projetoDTO.getProcessado());
                lkResultado.setOutputMarkupId(true);
                lkResultado.setOutputMarkupPlaceholderTag(true);
                item.add(lkResultado);
                projetoDTO.setLkResultado(lkResultado);

                AjaxCheckBox paraProcessarACB = new AjaxCheckBox("paraProcessarACB", new PropertyModel(projetoDTO, "paraProcessar")) {
                    @Override
                    protected void onUpdate(AjaxRequestTarget target) {

                        List<ProjetoDTO> listaProjetosProcessar = new ArrayList<>();

                        for (ProjetoDTO projetoDTO : listaProjetos) {
                            if (projetoDTO.getParaProcessar()) {
                                listaProjetosProcessar.add(projetoDTO);
                            }
                            ;
                        }

                        if (listaProjetosProcessar.size() > 0) {
                            processarTodos.setEnabled(true);
                        } else {
                            processarTodos.setEnabled(false);
                        }
                        target.add(processarTodos);
                    }
                };
                item.add(paraProcessarACB);

                item.add(new Label("nomeProjeto", projetoDTO.getName()));
                item.add(new Label("projeto", projetoDTO.getPath()));

                WebMarkupContainer iconProcessado = new WebMarkupContainer("iconProcessado");
                iconProcessado.setVisible(projetoDTO.getProcessado());
                iconProcessado.setOutputMarkupId(true);
                iconProcessado.setOutputMarkupPlaceholderTag(true);
                item.add(iconProcessado);
                projetoDTO.iconProcessado = iconProcessado;

                WebMarkupContainer iconNaoProcessado = new WebMarkupContainer("iconNaoProcessado");
                iconNaoProcessado.setVisible(!projetoDTO.getProcessado());
                iconNaoProcessado.setOutputMarkupId(true);
                iconNaoProcessado.setOutputMarkupPlaceholderTag(true);
                item.add(iconNaoProcessado);
                projetoDTO.iconNaoProcessado = iconNaoProcessado;

                WebMarkupContainer progressProject = new WebMarkupContainer("progressProject");
                progressProject.setOutputMarkupPlaceholderTag(true);
                progressProject.setOutputMarkupId(true);
                progressProject.add(new AttributeModifier("style", "width: " + projetoDTO.getProcentagem() + "%"));
                item.add(progressProject);
                projetoDTO.progressProject = progressProject;

                Label lbPorcetagem = new Label("lbPorcentagem", projetoDTO.getProcentagem());
                lbPorcetagem.setOutputMarkupId(true);
                lbPorcetagem.setOutputMarkupPlaceholderTag(true);
                projetoDTO.lbPorcentagem = lbPorcetagem;
                progressProject.add(lbPorcetagem);

            }
        };
        lvProjetos.setOutputMarkupId(true);
        lvProjetos.setOutputMarkupPlaceholderTag(true);
        add(lvProjetos);

    }

    /**
     * Creates a timer that periodically updates the UI with processing progress and logs.
     */
    private void criarTimer() {
        AbstractAjaxTimerBehavior timer = new AbstractAjaxTimerBehavior(java.time.Duration.ofSeconds(1)) {
            int cont = 0;

            @Override
            protected void onTimer(AjaxRequestTarget target) {
                taLog.setDefaultModel(Model.of(logRetorno));
                target.add(taLog);

                Boolean todosProjetosProcessados = true;

                List<ProjetoDTO> listaProjetosProcessar = new ArrayList<>();

                for (ProjetoDTO projetoDTO : listaProjetos) {

                    if (projetoDTO.getParaProcessar()) {
                        listaProjetosProcessar.add(projetoDTO);
                    }
                }

                for (ProjetoDTO projetoDTO : listaProjetosProcessar) {

                    if (!lkResultadoBotton.isEnabled()) {
                        lkResultadoBotton.setEnabled(projetoDTO.getProcessado());
                        target.add(lkResultadoBotton);
                    }


                    WebMarkupContainer lkResultado = projetoDTO.getLkResultado();
                    if (!lkResultado.isEnabled()) {
                        lkResultado.setEnabled(projetoDTO.getProcessado());
                        target.add(lkResultado);
                    }

                    WebMarkupContainer lkCharts = projetoDTO.lkCharts;
                    if (!lkCharts.isEnabled()) {
                        lkCharts.setEnabled(projetoDTO.getProcessado());
                        target.add(lkCharts);
                    }

                    WebMarkupContainer iconProcessado = projetoDTO.iconProcessado;
                    iconProcessado.setVisible(projetoDTO.getProcessado());
                    target.add(iconProcessado);

                    WebMarkupContainer iconNaoProcessado = projetoDTO.iconNaoProcessado;
                    iconNaoProcessado.setVisible(!projetoDTO.getProcessado());
                    target.add(iconNaoProcessado);

                    Label lbPorcentagem = projetoDTO.lbPorcentagem;
                    lbPorcentagem.setDefaultModel(Model.of(projetoDTO.getProcentagem()));
                    target.add(lbPorcentagem);

                    WebMarkupContainer progressProject = projetoDTO.progressProject;
                    progressProject.add(new AttributeModifier("style", "width: " + projetoDTO.getProcentagem() + "%"));
                    target.add(progressProject);

                    todosProjetosProcessados = todosProjetosProcessados && projetoDTO.getProcessado();
                }

                if (todosProjetosProcessados) {
                    totalProcessado.setValor((100 - totalProcessado.getValor()) + totalProcessado.getValor());
                    processando = false;
                }

                boolean processado = true;
                for (ProjetoDTO p : listaProjetosProcessar) {
                    processado = processado && p.getProcessado();
                }
            }
        };
        add(timer);
    }

}

File: src/main/java/io/github/arieslab/pages/ByClassTestPage.kt
package io.github.arieslab.pages

import io.github.arieslab.WicketApplication
import io.github.arieslab.base.JNose
import io.github.arieslab.base.Util
import io.github.arieslab.business.ProjetoBusiness
import io.github.arieslab.dtolocal.ProjetoDTO
import io.github.arieslab.dtolocal.TotalProcessado
import io.github.arieslab.entities.Projeto
import io.github.arieslab.pages.base.BasePage
import org.apache.wicket.AttributeModifier
import org.apache.wicket.ajax.AbstractAjaxTimerBehavior
import org.apache.wicket.ajax.AjaxRequestTarget
import org.apache.wicket.ajax.markup.html.form.AjaxCheckBox
import org.apache.wicket.extensions.ajax.markup.html.IndicatingAjaxLink
import org.apache.wicket.markup.html.WebMarkupContainer
import org.apache.wicket.markup.html.basic.Label
import org.apache.wicket.markup.html.link.Link
import org.apache.wicket.markup.html.list.ListItem
import org.apache.wicket.markup.html.list.ListView
import org.apache.wicket.markup.html.panel.FeedbackPanel
import org.apache.wicket.model.IModel
import org.apache.wicket.model.LambdaModel
import org.apache.wicket.model.Model
import org.apache.wicket.spring.injection.annot.SpringBean
import java.io.File
import java.time.Duration

class ByClassTestPage : BasePage("ByClassTestPage") {

    @SpringBean
    private lateinit var projetoBusiness: ProjetoBusiness

    private var pastaPath = ""
    private var pathAppToWebapp = BasePage.resolveRealPath()
    private var pastaPathReport = "$pathAppToWebapp${File.separatorChar}reports${File.separatorChar}"
    private val listaProjetos = mutableListOf<ProjetoDTO>()
    private lateinit var lvProjetos: ListView<ProjetoDTO>
    private lateinit var taLog: Label
    private val totalProcessado = TotalProcessado()
    private var processando = false
    private lateinit var processarTodos: IndicatingAjaxLink<String>
    private val logRetorno = StringBuffer()
    private var dataProcessamentoAtual = Util.dateNowFolder()
    private var processarCobertura = false
    private val listaResultado = mutableListOf<MutableList<String>>()
    private lateinit var lkResultadoBotton: Link<String>

    init {
        totalProcessado.setValor(0)

        criarCheckBoxCobertura()

        taLog = Label("taLog").apply {
            setEscapeModelStrings(false)
            outputMarkupId = true
            setOutputMarkupPlaceholderTag(true)
        }
        add(taLog)

        criarListaProjetos()

        add(FeedbackPanel("feedback").apply { outputMarkupId = true })

        criarBotaoProcessarTodos()

        lkResultadoBotton = object : Link<String>("lkResultado") {
            override fun onClick() {
                setResponsePage(ResultPage(listaResultado, "Result By ClassTest", "result_byclasstest", true))
            }
        }.apply {
            isEnabled = processando
            outputMarkupId = true
            setOutputMarkupPlaceholderTag(true)
        }
        add(lkResultadoBotton)

        add(object : Link<String>("lkCharts") {
            override fun onClick() {}
        }.apply { isVisible = false })

        loadProjetos()
        criarTimer()
    }

    private fun criarBotaoProcessarTodos() {
        processarTodos = object : IndicatingAjaxLink<String>("processarTodos") {
            override fun onClick(target: AjaxRequestTarget) {
                processando = true
                val listaParaProcessar = listaProjetos.filter { it.paraProcessar }
                val result = JNose.processarProjetos2(listaParaProcessar, dataProcessamentoAtual, totalProcessado, pastaPathReport, logRetorno)
                listaResultado.addAll(result)
            }
        }.apply { isEnabled = false }
        add(processarTodos)
    }

    private fun criarCheckBoxCobertura() {
        val acbCobertura = object : AjaxCheckBox("acbCobertura", Model.of(processarCobertura)) {
            override fun onUpdate(target: AjaxRequestTarget) {
                processarCobertura = !processarCobertura
                WicketApplication.COBERTURA_ON = processarCobertura
                println("COVERAGE_ON: $processarCobertura")
                logRetorno.insert(0, "COVERAGE_ON: $processarCobertura <br>")
            }
        }
        add(acbCobertura)
    }

    private fun loadProjetos() {
        dataProcessamentoAtual = Util.dateNowFolder()
        logRetorno.setLength(0)
        totalProcessado.setValor(0)
        val listaProjetosBean = projetoBusiness.listAllWithFilter()
        for (projeto in listaProjetosBean) {
            listaProjetos.add(ProjetoDTO.fromProjeto(projeto))
        }
        lvProjetos.list = listaProjetos
        processarTodos.isEnabled = true
    }

    private fun criarListaProjetos() {
        lvProjetos = object : ListView<ProjetoDTO>("lvProjetos", listaProjetos) {
            override fun populateItem(item: ListItem<ProjetoDTO>) {
                val projetoDTO = item.modelObject

                val lkCharts = object : Link<String>("lkCharts") {
                    override fun onClick() {}
                }.apply {
                    isEnabled = projetoDTO.processado
                    outputMarkupId = true
                    setOutputMarkupPlaceholderTag(true)
                }
                item.add(lkCharts)
                projetoDTO.lkCharts = lkCharts

                val lkResultado = object : Link<String>("lkResultado") {
                    override fun onClick() {
                        setResponsePage(ResultPage(projetoDTO.resultado!!, "Result By ClassTest: ${projetoDTO.name}", "${projetoDTO.name}_result_byclasstest", true, projetoDTO.coverageCsvPath))
                    }
                }.apply {
                    isEnabled = projetoDTO.processado
                    outputMarkupId = true
                    setOutputMarkupPlaceholderTag(true)
                }
                item.add(lkResultado)
                projetoDTO.lkResultado = lkResultado

                item.add(object : AjaxCheckBox("paraProcessarACB", Model.of(projetoDTO.paraProcessar)) {
                    override fun onUpdate(target: AjaxRequestTarget) {
                        projetoDTO.paraProcessar = !projetoDTO.paraProcessar
                        val listaProjetosProcessar = listaProjetos.filter { it.paraProcessar }
                        processarTodos.isEnabled = listaProjetosProcessar.isNotEmpty()
                        target.add(processarTodos)
                    }
                })

                item.add(Label("nomeProjeto", projetoDTO.name))
                item.add(Label("projeto", projetoDTO.path))

                val iconProcessado = WebMarkupContainer("iconProcessado").apply {
                    isVisible = projetoDTO.processado
                    outputMarkupId = true
                    setOutputMarkupPlaceholderTag(true)
                }
                item.add(iconProcessado)
                projetoDTO.iconProcessado = iconProcessado

                val iconNaoProcessado = WebMarkupContainer("iconNaoProcessado").apply {
                    isVisible = !projetoDTO.processado
                    outputMarkupId = true
                    setOutputMarkupPlaceholderTag(true)
                }
                item.add(iconNaoProcessado)
                projetoDTO.iconNaoProcessado = iconNaoProcessado

                val progressProject = WebMarkupContainer("progressProject").apply {
                    setOutputMarkupPlaceholderTag(true)
                    outputMarkupId = true
                    add(AttributeModifier("style", "width: ${projetoDTO.procentagem}%"))
                }
                item.add(progressProject)
                projetoDTO.progressProject = progressProject

                val lbPorcetagem = Label("lbPorcentagem", projetoDTO.procentagem).apply {
                    outputMarkupId = true
                    setOutputMarkupPlaceholderTag(true)
                }
                projetoDTO.lbPorcentagem = lbPorcetagem
                progressProject.add(lbPorcetagem)
            }
        }.apply {
            outputMarkupId = true
            setOutputMarkupPlaceholderTag(true)
        }
        add(lvProjetos)
    }

    private fun criarTimer() {
        add(object : AbstractAjaxTimerBehavior(Duration.ofSeconds(1)) {
            override fun onTimer(target: AjaxRequestTarget) {
                taLog.defaultModel = Model.of(logRetorno.toString())
                target.add(taLog)

                var todosProjetosProcessados = true
                val listaProjetosProcessar = listaProjetos.filter { it.paraProcessar }

                for (projetoDTO in listaProjetosProcessar) {
                    lkResultadoBotton.apply {
                        if (!isEnabled) {
                            isEnabled = projetoDTO.processado
                            target.add(this)
                        }
                    }

                    projetoDTO.lkResultado?.apply {
                        if (!isEnabled) {
                            isEnabled = projetoDTO.processado
                            target.add(this)
                        }
                    }

                    projetoDTO.lkCharts?.apply {
                        if (!isEnabled) {
                            isEnabled = projetoDTO.processado
                            target.add(this)
                        }
                    }

                    projetoDTO.iconProcessado?.apply {
                        isVisible = projetoDTO.processado
                        target.add(this)
                    }

                    projetoDTO.iconNaoProcessado?.apply {
                        isVisible = !projetoDTO.processado
                        target.add(this)
                    }

                    projetoDTO.lbPorcentagem?.apply {
                        defaultModel = Model.of(projetoDTO.procentagem)
                        target.add(this)
                    }

                    projetoDTO.progressProject?.apply {
                        add(AttributeModifier("style", "width: ${projetoDTO.procentagem}%"))
                        target.add(this)
                    }

                    todosProjetosProcessados = todosProjetosProcessados && projetoDTO.processado
                }

                if (todosProjetosProcessados) {
                    totalProcessado.setValor((100 - totalProcessado.valor) + totalProcessado.valor)
                    processando = false
                }
            }
        })
    }
}
